F2Py.cmake

```diff
@@ -38,21 +38,40 @@
 #                     These can be C files, Fortran files, object files. A single .pyf
 #                     file is also accepted.
 #
+# COMPILE_DEFS      : List of definitions for compiler. If not empty the list is past
+#                     to target_compile_definitions and to fortran_defs variable used
+#                     for preprocessing. The definitions are intended for preprocessing
+#                     of source files, i.e. it is implied that -D will prepended to each
+#                     definition
+#
 # The module generates a target with target_name. You can manipulate this target like
 # any other target in cmake to change its properties, for example, to set special
 # compiler flags.
 #
 # A log file is generated as a side effect with name <target>.log. The log is placed
 # in the current build directory.
+#
+# If the variable IMPY_DEV_PYF_GENERATION is defined (in impy/ImpyDevSettings.cmake),
+# new ${target_name}.pyf, ${target_name}module.c, ${target_name}-f2pywrappers.f will be
+# generated. Otherwise the old files from sources are used.
+
 
 function (f2py_add_module target_name)
 
   cmake_parse_arguments(F2PY_ADD_MODULE
     ""
     ""
-    "FUNCTIONS;INCLUDE_DIRS;INTERFACE_SOURCES;SOURCES"
+    "FUNCTIONS;INCLUDE_DIRS;INTERFACE_SOURCES;SOURCES;COMPILE_DEFS"
     ${ARGN})
 
+  if(NOT DEFINED ENV{IMPY_DEV_PYF_GENERATION})
+    list(PREPEND F2PY_ADD_MODULE_INTERFACE_SOURCES 
+      ${f2py_dir}/${target_name}.pyf
+      ${f2py_dir}/${target_name}module.c
+      ${f2py_dir}/${target_name}-f2pywrappers.f
+    ) 
+  endif()
+
```

Comment:
Also, if this env variable cannot be removed, do we really need such a long name? PYF_GENERATION should be unique enough or IMPY_PYF_GENERATION if you want to play it safe.

It is best to not only check whether a variable exists but also that its value can be interpreted as true.

Reply:
*.pyf files under git version control, so they should be changed rarely only in cases when development is finished and the version is established. Also it would be inconvenient to delete them by hand during development. So I think it is better not to copy automatically the newly generated *.pyf files. Having 2 version of files requires to have a switcher to choose what to use. This is IMPY_DEV_PYF_GENERATION in the current implementation.

There are 3 env variables `IMPY_DEV_PYF_GENERATION`, `IMPY_DEV_EXTRA_MODELS`,  `IMPY_DEV_DPMJETIII193`.  I would leave the prefix `IMPY_DEV` for consistency and to show that they belong to development mode.

> It is best to not only check whether a variable exists but also that its value can be interpreted as true.

Agree. It would be changed.

Reply:
> *.pyf files under git version control, so they should be changed rarely only in cases when development is finished and the version is established. Also it would be inconvenient to delete them by hand during development. So I think it is better not to copy automatically the newly generated *.pyf files. Having 2 version of files requires to have a switcher to choose what to use. This is IMPY_DEV_PYF_GENERATION in the current implementation.

And you don't need the switcher if files are only regenerated that do not exist. It is not at all inconvenient to delete a file that you want to be regenerated. It is quite natural. IMPY_DEV_PYF_GENERATION works on all files, while my proposal allows one to specifically recreate only the files for a specific model. If you want a one-liner to run during development, you can do
`rm -rf src/f2py/*.*; pip install -v .`

Reply:
I admit that this goes against what I said before that the build files should never override the source files, so I am open on this, but I would like to discuss it.

The question is what is more natural and what requires less code to implement.

Reply:
To clarify, my preferred solution is automatically generate pyf etc files when they do not exist in src/f2py, but to write those files to CMAKE_BINARY_DIR. It is then up to the developer to copy these files manually to src/f2py once development is finished. This satisfies all constrains.

Reply:
Ok. The working system is at [5f95def](https://github.com/impy-project/impy/pull/59/commits/5f95defbf9ff273d5cbd924f671c4c511ab41161)